simplemma/strategies/dictionaries/trie_directory_factory.py

```diff
@@ -0,0 +1,161 @@
+import hashlib
+import logging
+from collections.abc import MutableMapping
+from functools import lru_cache
+from pathlib import Path
+from typing import ByteString, Dict, List, Optional, cast
+
+from marisa_trie import BytesTrie, HUGE_CACHE  # type: ignore[import-not-found]
+from platformdirs import user_cache_dir
+
+from simplemma.strategies.dictionaries.dictionary_factory import (
+    SUPPORTED_LANGUAGES,
+    DictionaryFactory,
+    _load_dictionary_from_disk,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class TrieWrapDict(MutableMapping):
+    """Wrapper around BytesTrie to make them behave like dicts."""
+
+    def __init__(self, trie: BytesTrie):
+        self._trie = trie
+
+    def __getitem__(self, item):
+        return self._trie[item.decode()][0]
+
+    def __setitem__(self, key, value):
+        raise NotImplementedError
+
+    def __delitem__(self, key):
+        raise NotImplementedError
+
+    def __iter__(self):
+        for key in self._trie.iterkeys():
+            yield key.encode()
+
+    def __len__(self):
+        return len(self._trie)
+
+
+class TrieDictionaryFactory(DictionaryFactory):
+    """Memory optimized DictionaryFactory backed by MARISA-tries.
+
+    This dictionary factory creates dictionaries, which are backed by a
+    MARISA-trie instead of a dict, to make them consume very little
+    memory compared to the DefaultDictionaryFactory. Trade-offs are that
+    lookup performance isn't as good as with dicts.
+    """
+
+    __slots__: List[str] = []
+
+    def __init__(
+        self,
+        cache_max_size: int = 8,
```

Comment:
We are doing double caching: in-memory and in-disk.
These params make unclear what it is about.

Reply:
I completely agree. I just kept that parameter from `DefaultDictionaryFactory` to be able to use `TrieDictionaryFactory` as a drop-in replacement for it.